File: leetcode/python/Sort List/Sort_List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

from ListNode import ListNode
class Solution:
    # @param head, a ListNode
    # @return a ListNode
    def sortList(self, head):
        if head is None or head.next is None:
            return head
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        head1 = head
        head2 = slow.next
        slow.next = None
        head1 = self.sortList(head1)
        head2 = self.sortList(head2)
        head = self.merge(head1,head2)
        return head

    def merge(self,head1,head2):
        if head1 is None:
            return head2
        if head2 is None:
            return head1
        dump = ListNode(0)
        p = dump
        while head1 and head2:
            if head1.val < head2.val:
                p.next = head1
                head1 = head1.next
            else:
                p.next = head2
                head2 = head2.next
            p = p.next
        if head1:
            p.next = head1
        if head2:
            p.next = head2
        return dump.next

solu = Solution()
node1 = ListNode(3)
node2 = ListNode(2)
node3 = ListNode(4)
node1.next = node2
node2.next = node3
head = solu.sortList(node1)
while head:
    print head.val
    head = head.next

File: leetcode/cpp/SortList.cpp
//
// Created by zz on 2017/12/18.

// Time:  O(nlogn)
// Space: O(logn)

#include "ListNode.h"

class Solution {
public:
    ListNode *sortList(ListNode *head) {
        if (head || head->next) {
            return head;
        }

        ListNode *slow = head, *fast = head;
        while (fast->next && fast->next->next) {
            slow = slow->next;
            fast = fast->next->next;
        }

        fast = slow->next;
        slow->next = nullptr;
        return MergeTwoList(sortList(head), sortList(fast));
    }

private:
    ListNode *MergeTwoList(ListNode *l1, ListNode *l2) {
        ListNode dummy{0};
        ListNode *curr = &dummy;
        while (l1 && l2) {
            if (l1->val > l2->val) {
                curr->next = l2;
                l2 = l2->next;
            } else {
                curr->next = l1;
                l1 = l1->next;
            }
            curr = curr->next;
        }
        curr->next = l1 ? l1 : l2;
        return dummy.next;

    }

};